### DISCUSSION TITLE
Unresolved external symbol while compiling JSBSimForUnreal

### DISCUSSION THREAD

**Discussion by Informatyk1234** [JSBSim-Team/jsbsim#1109] [Support requests]
Hi there,
When I try to compile JSBSimForUnreal I have these errors (sorry for Polish language, "nierozpoznany symbol zewnętrzny means "unresolved external symbol")
![image](https://github.com/JSBSim-Team/jsbsim/assets/91948788/f44f57a7-2cf0-40c2-af38-4d21eef2a7a1)

I use visual studio 2022, but with Windows 10 SDK (10.0.18362.0) and MSVC v142 tools

**Comment by seanmcleod** [comment]
Those are references to some new classes that were added to JSBSim with this commit - https://github.com/JSBSim-Team/jsbsim/commit/e33e184f512449b8120769825b20e4d35df8ee51 recently. So it looks like you're compiling some portion of your Unreal application with the new header files but still linking to an old build of the JSBSim library. So you need to re-build the JSBSim library.

**Reply by Informatyk1234** [reply]
I re-built JSBSim library, but it didn't help

**Comment by seanmcleod** [comment]
Ah, I think I see the issue. The new classes aren't decorated with `JSBSIM_API` which is needed if you're linking to the JSBSim library as a dynamic library (DLL), which is the case when you're using JSBSim from an Unreal project. I'll submit a pull request today to fix this.

```c++
class JSBSIM_API FGFDMExec : public FGJSBBase
```

Where `JSBSIM_API` is defined as:

```c++
#if defined(_WIN32) && !defined(JSBSIM_STATIC_LINK)
  #ifdef JSBSIM_EXPORT
    #define JSBSIM_API __declspec(dllexport)
  #else
    #define JSBSIM_API __declspec(dllimport)
  #endif
#else
  #define JSBSIM_API
#endif
```

**Comment by seanmcleod** [comment]
I've just submitted a pull request - https://github.com/JSBSim-Team/jsbsim/pull/1111. You can either wait for it to be merged in or if you're in a rush just apply the handful of changes in the commit to your local source.

@bcoconni I guess we should add another build task to the JSBSim workflow. At a minimum to build the JSBSim library as a DLL and then build against it with some minimal host wrapper, or go all the way and build the Unreal project.

**Reply by bcoconni** [reply]
As far as building JSBSim as a DLL is concerned this is already done by the CI workflow (it is called `C/C++ build (MSVC) (ON)` and `C/C++ build (MinGW) (ON)`. The tag `(ON)` means that the property `shared_libs` is set to `ON` i.e. that JSBSim is built as a DLL.

https://github.com/JSBSim-Team/jsbsim/blob/c867cc7116283f08850db36444baf2577b967b67/.github/workflows/cpp-python-build.yml#L238-L239
https://github.com/JSBSim-Team/jsbsim/blob/c867cc7116283f08850db36444baf2577b967b67/.github/workflows/cpp-python-build.yml#L289-L290

Regarding the build of the Unreal plugin that's an entirely different story as I haven't found any example of a GitHub action for Unreal. And I'm afraid that it would be quite a challenge as the Unreal SDK is really huge. Unless someone can give some hints about how to have a minimal Unreal installation that would be sufficient to build JSBSim Unreal plugin.

**Comment by seanmcleod** [comment]
In terms of the size of the Unreal SDK etc. that's partly why I haven't got it installed locally to test etc. So when this report came in I was sort of guessing what the problem was without being able to test my guess.

I originally thought that the issue was that the new logging classes weren't being exported from the DLL version of JSBSim and that the Unreal project was trying to call them and so it failed during linking.

What I didn't realise at the time was that there is a separate Unreal specific project file `JSBSimForUnreal.vcxproj` that is used to build JSBSim, I assumed the Unreal project file was making use of a JSBSim DLL built by either CMake or `JSBSim.vcxproj`. I came across this when someone else reported the same problem, see - https://github.com/JSBSim-Team/jsbsim/issues/1115

So the issue wasn't the fact that the new logging classes weren't exported, the issue was that compiling JSBSim itself via `JSBSimForUnreal.vcxproj` was failing because the logging class .cpp file wasn't included in the build.
